doc/pins/send_to_ingress_hld.md

```diff
@@ -0,0 +1,178 @@
+
+Injecting packets to the ingress pipeline can also be used as a way to test ASIC routing behavior in the standalone test environment.
+
+## Requirements
+
+* Provide configuration support of send to ingress in the port table.
+
+## Architecture Design
+
+![drawing](images/send_to_ing.png)
+
+Details of the transmit path in GPINs Packet I/O can be found in this [HLD](https://github.com/sonic-net/SONiC/blob/master/doc/pins/Packet_io.md#transmit-path).
+
+## High-Level Design
+
+This HLD introduces a new “Send To Ingress” port type in the port table that SWSS will create a corresponding netdev/host interface. Any application can inject the packet to the ASIC’s ingress pipeline and let the ASIC decide which port to send packet out based on the routing table and current state of the network.
+
+Note that the Send to Ingress host interface is only for packet out as it is only used in TX path, no packet ins are expected.
+
+### CONFIG DB schema
+
+Add knob in PORT table in CONFIG DB or config_db.json to enable send to ingress
+
+```
+  "PORT": {
+    "Ethernet0": {
+      "admin_status": "up",
+      "alias": "NoAliasEth0/1",
+      "index": "1",
+      "lanes": "9,10,11,12,13,14,15,16",
+      "mtu": "9100",
+      "speed": "400000"
+    },
+    …
+  },
+  "SEND_TO_INGRESS_PORT": {
+     "SEND_TO_INGRESS": {}
+  },
+```
+
+### SWSS
+
+1. PortsMgr will process `SEND_TO_INGRESS_PORT|SEND_TO_INGRESS` in config_db.json during initialization, or add the `SEND_TO_INGRESS_PORT|SEND_TO_INGRESS` entry in CONFIG DB during runtime. It will be mapped to APPL_DB `SEND_TO_INGRESS_PORT_TABLE:SEND_TO_INGRESS` table entry.
+2. PortsOrch then picks up the entry in `SEND_TO_INGRESS_PORT_TABLE` in APPL DB and calls `sai_hostif_api->create_hostif` or `sai_hostif_api->delete_hostif` to  add/delete the send_to_ingress netdev.
+
+```
+sai_attribute_t attr;
+vector<sai_attribute_t> &ingress_attribs;
+
+attr.id = SAI_HOSTIF_ATTR_TYPE;
+attr.value.s32 = SAI_HOSTIF_TYPE_NETDEV;
+ingress_attribs.push_back(attr);
+attr.id = SAI_HOSTIF_ATTR_NAME;
+auto size = sizeof(attr.value.chardata);
+strncpy(attr.value.chardata, send_to_ingress_name.c_str(), size - 1);
+attr.value.chardata[size - 1] = '\0';
+ingress_attribs.push_back(attr);
+// If this isn't passed in true, the false setting makes
+// the device unready for later attempts to set UP/RUNNING
+attr.id = SAI_HOSTIF_ATTR_OPER_STATUS;
+attr.value.booldata = true;
+ingress_attribs.push_back(attr);
+// Get CPU port object id to signal send to ingress
+attr.id = SAI_SWITCH_ATTR_CPU_PORT;
```

Comment:
Why would CPU port be provided here? Is it used as an indication for SEND to ingress?

Reply:
Yes. P4Runtime will create a socket for the “send_to_ingress” netdev port and write into this socket for packets marked to send to “send_to_ingress”. Packets egressing out of the socket end up as incoming packets on the CPU port of the switch.

Ref https://github.com/sonic-net/SONiC/blob/master/doc/pins/Packet_io.md#transmit-path-design